doc/design/proposal-synchronous-hooks.md

```diff
@@ -0,0 +1,258 @@
+    return { exports: newExports }
+  }
+  return {
+    exports: originalExports
+  };
+}
+```
+
+It's not yet investigated whether it is possible to make it work with the CJS loader cache at all, but it looks closer to what developers generally try to monkey-patch the CJS loader for.
+
+## `link` (import-only): invoked before linking
+
+This is invoked after `load` but prior to Node.js passing the final result to V8 for linking, so it can be composed with `resolve` and `load` if necessary. This needs to work with experimental vm modules for passing module instances around.
+
+```js
+/**
+ * @typedef {{
+ *  source: string | Buffer
+ * }} ModuleLinkContext
+ */
+/**
+ * @typedef {{
+*   module: vm.Module
+* }} ModuleLinkResult
+*/
+/**
+ * @param {string} url
+ * @param {ModuleLinkContext} context
+ * @param {(context: ModuleLinkContext) => {ModuleLinkResult}} nextLink
+ * @returns {ModuleLinkResult}
+ */
+function link(url, context, nextLink) {  // Mocking import-in-the-middle
+  const { module: originalModule } = nextLink(url, context);
+  assert.strictEqual(module.status, 'linked');  // Original module is linked at this point
+  let source = `import * as original from 'original';`;
+  source += `import { userCallback, name, basedir } from 'util'`;
+  source += `const exported = {}`;
+  for (const key of originalModule.namespace) {
+    source += `let $${key} = original.${key};`;
+    source += `export { $${key} as ${key} }`;
+    source += `Object.defineProperty(exported, '${key}', { get() { return $${key}; }, set (value) { $${key} = value; }});`;
+  }
+  source += `userCallback(exported, name, basedir);`;
+  const m = vm.SourceTextModule(source);
+  m.linkSync((specifier) => {  // This is not yet implemented but should be trivial to implement.
+    if (specifier === 'original') return originalModule;
+    // Contains a synthetic module with userCallback, name & basedir computed from url
+    if (specifier === 'util') return util;
+  });
+  return { module: m };
+}
+```
+
+### Alternative design: `link` that encompasses `resolve` and `load`
+
+An alternative design would be, instead of invoking a hook after `load`, make it wrap around `resolve` and `load` steps. The `link` hooks would take `specifier` and return `vm.Module` instances. If `nextLink()` is invoked, the next `resolve` and `load` (e.g. the default ones) will be invoked inside that. If `nextLink()` is not invoked, the default `resolve` and `load` will be skipped.
+
+```js
+function link(specifier, context, nextImport) {  // Mocking import-in-the-middle
+  const { module: originalModule } = nextLink(specifier, context);
+  assert.strictEqual(module.status, 'linked');  // Original module is linked at this point
+  let source = '...';
+  const m = vm.SourceTextModule(source);
+  m.linkSync((specifier) => {
+    // ...
+  });
+  return { module: m };
+}
+```
+
+## Other use cases that may need some thought
+
+1. Hot module replacement
```

Comment:
Currently, hmr is one of the reason to use `require.cache`. I think if we want to change the loader API, we really need to have a good solution for hot module reloading.

Reply:
The `require.cache` is also used extensively in testing. Particularly in APM tests. We routinely want to wipe out what was already loaded for one test so that we can get a clean slate for a next test.

Reply:
I think what we need to figure out is whether we want to make cache modification capability a blocker of adding hooks that specify what should be done when a module is reloaded, or wether it's fine to just have that in mind and design the hooks in a way that won't prevent cache modification (at least with the current proposal and ideas I don't see how they are in conflict - there can be a separate API for modifying caches, but it's not in conflict with the hooks here); my hunch is if the two are coupled then we might have neither for a very long time, while monkey-patching continues to get relied on.

Reply:
Added a section about cache manipulation - I think so far the patterns I have found are like this:

```js
// Maybe wrapped by some fs.watch, or overriding the default require somehow
delete require.cache[require.resolve(specifier)];
```

I think the manipulation can probably be its own API that can be composed with the hooks which are run when a module gets loaded for the first time or reloaded after the cache is cleared. At least I am not seeing too much conflict with the current proposal.